File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/242-valid-anagram-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function isAnagram(s: string, t: string): boolean {
    if (s.length !== t.length) return false;
    const counts = new Map<string, number>();
    for (const ch of s) counts.set(ch, (counts.get(ch) ?? 0) + 1);
    for (const ch of t) {
        const c = counts.get(ch);
        if (!c) return false;
        counts.set(ch, c - 1);
    }
    return true;
}

assert(isAnagram('anagram', 'nagaram') === true);
assert(isAnagram('rat', 'car') === false);
assert(isAnagram('a', 'a') === true);
assert(isAnagram('ab', 'ba') === true);
assert(isAnagram('ab', 'a') === false);
assert(isAnagram('', '') === true);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/242-valid-anagram-approach2.py
from collections import Counter

def is_anagram(s: str, t: str) -> bool:
    return Counter(s) == Counter(t)

assert is_anagram('anagram', 'nagaram') == True
assert is_anagram('rat', 'car') == False
assert is_anagram('a', 'a') == True
assert is_anagram('ab', 'ba') == True
assert is_anagram('ab', 'a') == False
assert is_anagram('', '') == True
print("all tests pass")
